### DISCUSSION TITLE
Should graphblas.matrix_select simply be graphblas.select?

### DISCUSSION THREAD

**Discussion by paul-tqh-nguyen** [metagraph-dev/mlir-graphblas#194] [Ideas]
Should `graphblas.matrix_select` simply be `graphblas.select`?

It would make sense that the verifier would enforce that `tril` and `triu` only work with matrices, but for a selector like `gt`, it would make sense for this selector to work with vectors. 

Also, should we merge `graphblas.matrix_select_random` into `graphblas.matrix_select`? Or is it sufficiently different? If it is sufficiently different, then what makes (`graphblas.matrix_select`+`gt`) not sufficiently different than (`graphblas.matrix_select`+`tril`)?

**Comment by seibert** [comment]
Yes, I think it would be good to remove terms like `matrix` and `vector` from ops whenever the semantics are clear, especially since we know the types of the inputs at code generation type and can switch between vector and matrix specific implementations very easily.

`select_random` is very different than `select`, though, because the selection condition is not independent for every element.  (Something like `gt0` or `tril` can be evaluated in parallel on every element if you want to, for example.)  In fact, I think `select_random` is not a good name for it, but I'm not sure what to call it.  It is more of a `select_random_from_rows`, so it can process each row in a CSR independently, but within the row the selection is not independent (you want exactly N elements, with no duplication).

**Comment by jim22k** [comment]
In `grblas`, we have:
- `select` (for vector or matrix)
- `selectk` (for vector, similar to select_random)
- `selectk_rowwise` (for matrix, same as the current matrix_select_random)
- `selectk_colwise` (for matrix, does the same but along columns)

I think we should do something similar in mlir-graphblas. We should also group our selectors based on whether they compare:
- rows & cols (only applies to matrix)
- index (only applies to vector)
- value (applies to both matrix and vector)

Also, I would like to propose that all selectors take a `thunk`. Even something like `triu` can take a thunk indicating how far off-diagonal the selection is limited to. This would make it much easier to handle multiple selectors in the dialect signature.
